atc/db/watch/list_all_jobs_watcher_test.go

```diff
@@ -0,0 +1,364 @@
+			build, _ := job1.CreateBuild()
+			Eventually(getInvokedEvents).Should(HaveLen(1))
+			resetInvokedEvents()
+
+			build.Finish(db.BuildStatusSucceeded)
+			Eventually(getInvokedEvents).Should(ConsistOf(PutEvent(job1)))
+		})
+
+		It("notifies on pipeline renamed", func() {
+			pipeline.Rename("blah")
+			Eventually(getInvokedEvents).Should(ConsistOf(PutEvent(job1), PutEvent(job2)))
+		})
+
+		It("notifies on team renamed", func() {
+			team.Rename("blah")
+			Eventually(getInvokedEvents).Should(ConsistOf(PutEvent(job1), PutEvent(job2)))
+		})
+
+		It("notifies on team renamed", func() {
+			team.Rename("blah")
+			Eventually(getInvokedEvents).Should(ConsistOf(PutEvent(job1), PutEvent(job2)))
+		})
+
+		It("notifies on job deactivated", func() {
+			team.SavePipeline("pipeline", atc.Config{
+				Jobs: atc.JobConfigs{{Name: "job2"}},
+			}, pipeline.ConfigVersion(), false)
+
+			Eventually(getInvokedEvents).Should(ConsistOf(DeleteEvent(job1.ID()), PutEvent(job2)))
+		})
+
+		It("notifies on job reactivated", func() {
+			pipeline, _, _ = team.SavePipeline("pipeline", atc.Config{
+				Jobs: atc.JobConfigs{{Name: "job2"}},
+			}, pipeline.ConfigVersion(), false)
+
+			Eventually(getInvokedEvents).Should(HaveLen(2))
+			resetInvokedEvents()
+
+			team.SavePipeline("pipeline", atc.Config{
+				Jobs: atc.JobConfigs{{Name: "job1"}, {Name: "job2"}},
+			}, pipeline.ConfigVersion(), false)
+			Eventually(getInvokedEvents).Should(ConsistOf(PutEvent(job1), PutEvent(job2)))
+		})
+
+		It("notifies on job deleted", func() {
+			pipeline.Destroy()
+
+			Eventually(getInvokedEvents).Should(ConsistOf(DeleteEvent(job1.ID()), DeleteEvent(job2.ID())))
+		})
+
+		It("notifies on pipeline exposed", func() {
+			pipeline.Expose()
+
+			Eventually(getInvokedEvents).Should(ConsistOf(PutEvent(job1), PutEvent(job2)))
+		})
+
+		It("notifies on pipeline hidden", func() {
+			pipeline.Hide()
+
+			Eventually(getInvokedEvents).Should(ConsistOf(PutEvent(job1), PutEvent(job2)))
+		})
+
+		Describe("access control for non-admins", func() {
+			BeforeEach(func() {
+				access.IsAdminReturns(false)
+			})
+
+			Context("when the user has access to the team", func() {
+				BeforeEach(func() {
+					access.TeamNamesReturns([]string{"team"})
+				})
+
+				It("forwards the notification", func() {
+					job1.CreateBuild()
+					Eventually(getInvokedEvents).Should(ContainElement(PutEvent(job1)))
+				})
+			})
+
+			Context("when the user does not have access to the team", func() {
+				BeforeEach(func() {
+					access.TeamNamesReturns([]string{"other-team"})
+				})
+
+				It("does not forward the notification", func() {
+					job1.CreateBuild()
+					Consistently(getInvokedEvents).Should(BeEmpty())
+				})
+
+				Context("when the notification is a delete", func() {
+					It("forwards the notification", func() {
+						pipeline.Destroy()
+						Eventually(getInvokedEvents).Should(ContainElement(DeleteEvent(job1.ID())))
+					})
+				})
```

Comment:
What's the scenario where I'm not part of the team but should be notified a job is deleted?

Reply:
Hmm yeah, this is really more of a limitation of the approach than a desired feature, so maybe this test shouldn't exist.

If we wanted to avoid sending notifications when jobs that aren't visible to a user are deleted, it would end up being pretty costly (in terms of additional queries and triggers) and probably pretty complicated - so I figured the tradeoff of sending every delete event to all clients was worth it (especially given that there's no sensitive info on delete events - just the job ID)